src/tqec/computation/open_graph.py

```diff
@@ -0,0 +1,220 @@
+"""Utilities for working with graphs with open ports."""
+
+from dataclasses import dataclass
+from functools import reduce
+from itertools import combinations
+from typing import Iterator
+
+import stim
+import networkx as nx
+
+from tqec.computation.block_graph import BlockGraph
+from tqec.computation.correlation import CorrelationSurface
+from tqec.computation.cube import YHalfCube, ZXCube
+from tqec.utils.enums import Basis
+from tqec.utils.exceptions import TQECException
+from tqec.utils.position import Direction3D
+
+
+@dataclass(frozen=True)
+class FilledGraph:
+    """A block graph without open ports that can be used for simulation.
+
+    Attributes:
+        graph: The block graph with all ports filled.
```

Comment:
That should be checked in the `__post_init__` method along with the other class invariants (e.g., the provided `stabilizers` are valid).